java/src/main/java/ai/rapids/cudf/DType.java

```diff
@@ -17,76 +17,202 @@
 
 import java.util.EnumSet;
 
-public enum DType {
-  EMPTY(0, 0, "NOT SUPPORTED"),
-  INT8(1, 1, "byte"),
-  INT16(2, 2, "short"),
-  INT32(4, 3, "int"),
-  INT64(8, 4, "long"),
-  UINT8(1, 5, "uint8"),
-  UINT16(2, 6, "uint16"),
-  UINT32(4, 7, "uint32"),
-  UINT64(8, 8, "uint64"),
-  FLOAT32(4, 9, "float"),
-  FLOAT64(8, 10, "double"),
-  /**
-   * Byte wise true non-0/false 0.  In general true will be 1.
-   */
-  BOOL8(1, 11, "bool"),
-  /**
-   * Days since the UNIX epoch
-   */
-  TIMESTAMP_DAYS(4, 12, "date32"),
+public final class DType {
+
+  public enum DTypeEnum {
+    EMPTY(0, 0, "NOT SUPPORTED"),
+    INT8(1, 1, "byte"),
+    INT16(2, 2, "short"),
+    INT32(4, 3, "int"),
+    INT64(8, 4, "long"),
+    UINT8(1, 5, "uint8"),
+    UINT16(2, 6, "uint16"),
+    UINT32(4, 7, "uint32"),
+    UINT64(8, 8, "uint64"),
+    FLOAT32(4, 9, "float"),
+    FLOAT64(8, 10, "double"),
+    /**
+     * Byte wise true non-0/false 0.  In general true will be 1.
+     */
+    BOOL8(1, 11, "bool"),
+    /**
+     * Days since the UNIX epoch
+     */
+    TIMESTAMP_DAYS(4, 12, "date32"),
+    /**
+     * s since the UNIX epoch
+     */
+    TIMESTAMP_SECONDS(8, 13, "timestamp[s]"),
+    /**
+     * ms since the UNIX epoch
+     */
+    TIMESTAMP_MILLISECONDS(8, 14, "timestamp[ms]"),
+    /**
+     * microseconds since the UNIX epoch
+     */
+    TIMESTAMP_MICROSECONDS(8, 15, "timestamp[us]"),
+    /**
+     * ns since the UNIX epoch
+     */
+    TIMESTAMP_NANOSECONDS(8, 16, "timestamp[ns]"),
+
+    //We currently don't have mappings for duration type to I/O files, and these
+    //simpleNames might change in future when we do
+    DURATION_DAYS(4, 17, "int32"),
+    DURATION_SECONDS(8, 18, "int64"),
+    DURATION_MILLISECONDS(8, 19, "int64"),
+    DURATION_MICROSECONDS(8, 20, "int64"),
+    DURATION_NANOSECONDS(8, 21, "int64"),
+    //DICTIONARY32(4, 22, "NO IDEA"),
+
+    STRING(0, 23, "str"),
+    LIST(0, 24, "list"),
+    DECIMAL32(4, 25, "decimal32"),
+    DECIMAL64(8, 26, "decimal64"),
+    STRUCT(0, 27, "struct");
+
+    private static final DTypeEnum[] D_TYPES = DTypeEnum.values();
+    final int sizeInBytes;
+    final int nativeId;
+    final String simpleName;
+
+    DTypeEnum(int sizeInBytes, int nativeId, String simpleName) {
+      this.sizeInBytes = sizeInBytes;
+      this.nativeId = nativeId;
+      this.simpleName = simpleName;
+    }
+  }
+
+  final DTypeEnum typeId;
+  final int scale;
+
+  private DType(DTypeEnum id) {
+    typeId = id;
+    scale = 0;
+  }
+
+  // TODO : Java docs for constructor and public methods
   /**
-   * s since the UNIX epoch
+   * Constructor for Decimal Type
+   * @param id Enum representing data type.
+   * @param decimalScale scale - number of digits to the right of the decimal point in a number
    */
-  TIMESTAMP_SECONDS(8, 13, "timestamp[s]"),
+  public DType(DTypeEnum id, int decimalScale) {
+    typeId = id;
```

Comment:
This needs to throw if the type is not a decimal type, otherwise the caller can create an instance of something like `DType.INT32` and now it's no longer a singleton, breaking `==` and `!=` checks on it.  The documentation for this constructor should be updated to reflect the behavior.

Reply:
This wasn't a private constructor when I first raised the issue.  😉 

Now that this is moved to being callable only from a factory and the factory has the checks, I'm happy.